---
title: "1_Love_prepare"
author: "Jeroen Gilis"
date: "06/11/2020"
output: html_document
---

**In order to run this script (1_Love_prepare.Rmd), the folder Love_salmon.zip, which contains the salmon quantification files for this dataset, should be downloaded from Zenodo and unzipped. In addition, the metaData files Love_metadata.rda should also be downloaded from Zenodo. Both files must then be copied into the Data folder of this Github page.** Note that the quantification files were generated from the same reads as those from the paper by Love et al.[swimming downstream](https://doi.org/10.12688/f1000research.15398.3). See the three Zenodo links from that paper for access to the raw data. However, in our publication, we used a more recent version of salmon (salmon v1.1.0) for quantification. Our provided metadata file is an exact copy of the simulate.rda file from the Love et al. dataset (see their Zenodo link).

**If you do not want to run this script, its outputs can also be downloaded from Zenodo: Love_benchmark_datasets_scaledTPM.Rdata (or, alernatively, Love_benchmark_datasets_count.Rdata) and Love_eff_len.Rds**

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FASLE}
library(tximport)
library(DRIMSeq)
library(edgeR)
library(doMC)
library(BANDITS)
```

# Load data

```{r}
root.dir <- rprojroot::find_rstudio_root_file()

sample_names <- c("sample_01","sample_02","sample_03","sample_04","sample_05","sample_06","sample_07","sample_08","sample_09","sample_10","sample_11","sample_12","sample_13","sample_14","sample_15","sample_16","sample_17","sample_18","sample_19","sample_20","sample_21","sample_22","sample_23","sample_24")
quant_files <- file.path(root.dir, "Data/Love_salmon", sample_names, "quant.sf")
file.exists(quant_files)

txi <- tximport(files = quant_files, type = "salmon", countsFromAbundance = "no", txOut = TRUE)
#txi <- tximport(files = quant_files, type = "salmon", countsFromAbundance = "scaledTPM", txOut = TRUE)

myDesign <- data.frame(sample_id = as.character(sample_names),
                       condition = as.character(rep(c("a","b"),each=12)))

colnames(txi$counts) <- myDesign$sample_id
```

# Load metadata


```{r}
load("./Data/Love_metadata.rda")

### Make our tx info copy (to match other benchmark data)
txInfo <- txdf[,2:1]
colnames(txInfo) <- c('isoform_id','gene_id')

txdf <- txdf[match(rownames(tpms), txdf$TXNAME),]
txdf$dtu.genes <- iso.dtu | iso.dte & !iso.dte.only
full.dtu.genes <- unique(txdf$GENEID[txdf$dtu.genes])

txp.exprs <- rowSums(tpms) > 0
dtu.dte.genes <- unique(txdf$GENEID[iso.dte & !iso.dte.only])
txdf$full.dtu <- iso.dtu | (txdf$GENEID %in% dtu.dte.genes & txp.exprs)
dtu.txps <- txdf$TXNAME[txdf$full.dtu] # Used as the truth in Love et al 2018

### Assign truth to our info df
txInfo$geneModified <- txInfo$isoform_id %in% dtu.txps
```

# Compute effective transcript lengths for BANDITS

```{r}
eff_len <- eff_len_compute(x_eff_len = txi$length)
saveRDS(eff_len, "./Data/Love_eff_len.Rds")

equiv_classes_files <- file.path(root.dir, "Data/Love_salmon", sample_names, "aux_info", "eq_classes.txt")
all(file.exists(equiv_classes_files))
```

# Setup generation of benchmark data

Set hyperparameters of the benchmark datasets (i.e. sample size, number of repeats, fraction of DTU genes and parameters for parallel processing).

```{r}
### Set parameters
samplesPrCondition   <- c(3,6,10)
nrRepsMade           <- 3
nrCoresToUse         <- 2

### Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'

    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

### list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
```

# Lenient filtering

Note; I do not used the getBenchmark_data.R script here to generate the benchmark datasets as the function deefined in that script still

```{r}
LoveBenchmarkLenient <- lapply(c(1:length(nrRepList)), function(x) {
      
    ### Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]
      
    if(TRUE) {
          
        samplesToUseCond1 <- sample(myDesign$sample_id[which(myDesign$condition == 'a')], localSampleSize)
        samplesToUseCond2 <- sample(myDesign$sample_id[which(myDesign$condition == 'b')], localSampleSize)
        
        samplesToUseCond1 <- as.character(samplesToUseCond1)
        samplesToUseCond2 <- as.character(samplesToUseCond2)

        localDesign <- myDesign[which(
            myDesign$sample_id %in% c(
                samplesToUseCond1,
                samplesToUseCond2
            )
        ),]
        }

        ### Step 2: Subset to expressed features using edgeR::filterByExpr
        if(TRUE) {
          
            y <- edgeR::DGEList(counts = txi$counts[,localDesign$sample_id])
            design <- model.matrix(~condition, data=localDesign)
            
            filter <- edgeR::filterByExpr(y,design=design) 
            
            localCm <- y$counts[filter,]
            
            ## Get only multi-isoform genes (after filtering)
            localTx <- txInfo[which(
                txInfo$isoform_id %in% rownames(localCm)),]
            
            tmp <- table(localTx$gene_id)
            tmp <- tmp[which( tmp >= 2)]
        
            localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
            localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
            
            ## add column with sample size for easy retrieval
            localTx$nrSamplesPerCondition <- localSampleSize
        }

        colnames(localTx) <- c('TXNAME','GENEID','gene_modified','nrSamplesPerCondition')

        ### Combine data
        dataList <- list(
            data     = localCm,
            design   = localDesign,
            metaInfo = localTx
        )
        
        return(dataList)
    }
)

names(LoveBenchmarkLenient) <- paste0(names(nrRepList),"_filterLenient")
```

# Stringent filtering

```{r}
LoveBenchmarkStringent <- lapply(c(1:length(nrRepList)), function(x) {
      
    ### Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]
      
    if(TRUE) {
          
        samplesToUseCond1 <- sample( myDesign$sample_id[which(myDesign$condition == 'a')], localSampleSize)
        samplesToUseCond2 <- sample( myDesign$sample_id[which(myDesign$condition == 'b')], localSampleSize)
        
        samplesToUseCond1 <- as.character(samplesToUseCond1)
        samplesToUseCond2 <- as.character(samplesToUseCond2)

        localDesign <- myDesign[which(
            myDesign$sample_id %in% c(
                samplesToUseCond1,
                samplesToUseCond2
            )
        ),]
    }
    
    

        ### Step 2: Subset to expressed features using DRIMSeq::dmFilter
        if(TRUE) {
          
          localCm <- txi$counts[,localDesign$sample_id]
          localCm <- as.data.frame(localCm)
          colnames(localCm)

          geneForEachTx <- txInfo[match(rownames(localCm),txInfo[,"isoform_id"]),"gene_id"]

          localCm$gene_id <- geneForEachTx
          localCm$feature_id <- row.names(localCm)
          

          d <- DRIMSeq::dmDSdata(counts = localCm, samples = localDesign)

          d <- dmFilter(d,
                min_samps_feature_expr=localSampleSize/2, min_feature_expr=10, min_samps_feature_prop=localSampleSize/2, min_feature_prop=0.1,
                min_samps_gene_expr=localSampleSize, min_gene_expr=10)

          localCm <- txi$counts[counts(d)$feature_id,localDesign$sample_id]

          ## Get only multi-isoform genes (after filtering)
          localTx <- txInfo[which(
              txInfo$isoform_id %in% rownames(localCm)),]
          
          tmp <- table(localTx$gene_id)
          tmp <- tmp[which( tmp >= 2)]
        
          localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
          localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
            
            
          ## add column with sample size for easy retrieval
          localTx$nrSamplesPerCondition <- localSampleSize
        }

        colnames(localTx) <- c('TXNAME','GENEID','gene_modified','nrSamplesPerCondition')

        ### Combine data
        dataList <- list(
            data     = localCm,
            design   = localDesign,
            metaInfo = localTx
        )

        return(dataList)
    }
)

names(LoveBenchmarkStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save Love benchmark data

```{r}
save(LoveBenchmarkLenient, LoveBenchmarkStringent, file="./Data/Love_benchmark_datasets_count.Rdata")
#save(LoveBenchmarkLenient, LoveBenchmarkStringent, file="./Data/Love_benchmark_datasets_scaledTPM.Rdata")
```




