---
title: "edgeR_DGE_analysis"
author: "Jeroen Gilis"
date: "11/11/2020"
output: html_document
---

This is scripts performs a DGE analysis with edgeR on a subset of cells from the single-cell RNA-Seq dataset by Tasic et al. [Tasic paper](https://doi.org/10.1038/s41586-018-0654-5). In order to run this script, three files should be downloaded from Zenodo and placed in the data folder of this GitHub repository;

1. Tasic_caseStudy_transcript_counts.Rds (transcript-level expression matrix)
2. Tasic_metadata_1.xlsx
3. Tasic_metadata_2.csv

Note that the metadata files are identical to the one supplemented by Tasic et al., Supplementary_Table_10_Full_Metadata.xlsx and GSE115746_accession_table.csv, respectively. All figures generated in this script are available from our GitHub repository under ./Results/CaseStudy/

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE}
## for analysis
library(edgeR)
library(tidyverse)
```

# Data import

## Import counts

```{r}
Tasic_gene_counts <- readRDS("./Data/Tasic_caseStudy_gene_counts.Rds")
# Already remove genes with zero counts in all cells
Tasic_gene_counts <- Tasic_gene_counts[rowSums(Tasic_gene_counts) != 0,]
dim(Tasic_gene_counts)
```

## Import metadata

```{r}
access <- read.csv2("./Data/Tasic_metadata_2.csv",sep = "\t")
access <- access[match(colnames(Tasic_gene_counts),access$SRA_Run),]

metaData <- openxlsx::read.xlsx("./Data/Tasic_metadata_1.xlsx")
metaData <- metaData[match(access$sample_name,metaData$sample_name),]
colnames(Tasic_gene_counts) <- metaData$sample_name
metaData <- metaData[metaData$core_intermediate_call == "Core",] # only retain cell that were unambiguously assigned to a certain cell type (cluster)
Tasic_gene_counts <- Tasic_gene_counts[,metaData$sample_name]

metaData <- metaData[,c("sample_name", "brain_region","cluster")]
```

# Set up the experimental design

```{r}
metaData$cluster <- gsub(" ", "_", metaData$cluster)
metaData$group <- paste(metaData$brain_region,metaData$cluster,sep=".")

# Remove groups (cell types) with less than 30 cells --> removes 4 groups
remove <- names(table(interaction(metaData$brain_region,metaData$cluster))[table(interaction(metaData$brain_region,metaData$cluster)) < 30])

metaData <- metaData[-which(metaData$group %in% remove),]
group <- as.factor(metaData$group)
 
design <- model.matrix(~ 0 + group) # Factorial design
colnames(design) <- levels(group)

Tasic_gene_counts <- Tasic_gene_counts[,which(colnames(Tasic_gene_counts) %in% metaData$sample_name)]
```

# Filtering

```{r}
filter_all_edgeR_stringent <- filterByExpr(Tasic_gene_counts,
             design = NULL,
             group = metaData$brain_region,
             lib.size = NULL,
             min.count = 10,
             min.total.count = 0,
             large.n = 0,
             min.prop = 0.7)

table(filter_all_edgeR_stringent)
Tasic_gene_counts <- Tasic_gene_counts[filter_all_edgeR_stringent,]
```

# Visualization: Generate Figure 6 panel D

DGE figure for the ENSMUSG00000029470 gene in contrast 6, i.e. VISp Hsd11b1-Endou cells versus ALM Tnc cells 

```{r}
# prepare for plot
plot_data <- cbind(metaData[metaData$group %in% c("ALM.L5_IT_ALM_Tnc","VISp.L5_IT_VISp_Hsd11b1_Endou"),],Tasic_gene_counts["ENSMUSG00000029470",metaData$group %in% c("ALM.L5_IT_ALM_Tnc","VISp.L5_IT_VISp_Hsd11b1_Endou")]) ## sample order is the same for both dataframes so we can do this
colnames(plot_data)[5] <- "counts"

plot_data$group[which(plot_data$group == "VISp.L5_IT_VISp_Hsd11b1_Endou")] <- "Hsd11b1_Endou"
plot_data$group[which(plot_data$group == "ALM.L5_IT_ALM_Tnc")] <- "Tnc"
plot_data$group <- factor(plot_data$group, levels=c("Tnc", "Hsd11b1_Endou"))

plot_data$variable <- "ENSMUSG00000029470"

label_facet <- function(geneID){
  lev <- levels(as.factor(geneID))
  lab <- "edgeR FDR = 1"
  names(lab) <- lev
  return(lab)
}

# plot
png("./Results/caseStudy/Figure6D.png",
    width     = 5,
    height    = 5,
    units     = "in",
    res       = 200,
    pointsize = 4)
ggplot(plot_data, mapping = aes(x=group,y=counts,fill=brain_region)) +
    geom_violin() +
    scale_fill_manual(values=c("royalblue4", "firebrick")) +
    geom_point(data = plot_data, aes(x=group,y=counts),size=0.6, position = position_jitterdodge(jitter.width = 0.7,jitter.height = 0, dodge.width = 0.9)) +
    stat_summary(fun=mean, geom="point", position = position_dodge(width = 0.9), shape=18, size=4, colour = "cyan") + 
    stat_summary(fun=median, geom="point", position = position_dodge(width = 0.9), shape=18, size=3, colour = "green4") +
    theme_bw() +
    ggtitle("ENSMUSG00000029470") +
    facet_wrap(~variable, ncol=1, labeller = labeller(variable = label_facet(plot_data$variable))) +
    ylab("gene expression count") +
    xlab("group") +
    theme(axis.text.x = element_text(angle = 0, vjust = 1, hjust=0.5, size=9)) +
    theme(strip.text = element_text(size = 9,face="bold")) +
    theme(plot.title = element_text(size = 9.5,face="bold"))
dev.off()

#plot_data$counts_scaled <- plot_data$counts/y$samples$norm.factors[metaData$group %in% c("ALM.L5_IT_ALM_Tnc","VISp.L5_IT_VISp_Hsd11b1_Endou")]
```

# DGE analysis with edgeR

## Fit models

```{r}
y <- DGEList(counts=Tasic_gene_counts, group=group)
y <- calcNormFactors(y)
y <- estimateDisp(y,design = design) # takes about 4 minutes
fit <- glmQLFit(y, design =  design) # fast
```

## Construct design matrix

```{r}
# Create same contrast matrix as for DTU analysis (same as original Tasic et al. analysis)
L <- matrix(0, ncol = 8, nrow = ncol(design))
rownames(L) <- colnames(design)
colnames(L) <- c("C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")

L[c("VISp.L5_IT_VISp_Batf3", "ALM.L5_IT_ALM_Cpa6_Gpr88"), 1] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col27a1", "ALM.L5_IT_ALM_Cbln4_Fezf2"), 2] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col6a1_Fezf2", "ALM.L5_IT_ALM_Cpa6_Gpr88"), 3] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col6a1_Fezf2", "ALM.L5_IT_ALM_Gkn1_Pcdh19"), 4] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Lypd1_Gpr88"), 5] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Tnc"), 6] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Tmem163_Dmrtb1"), 7] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Whrn_Tox2", "ALM.L5_IT_ALM_Tmem163_Arhgap25"), 8] <- c(1,-1)
```

## test

```{r}
# with TREAT (~fold-change threshold)
resultsTreat <- apply(L, 2, function(i) glmTreat(glmfit=fit, contrast=i, lfc=1))
topTreat <- lapply(resultsTreat, topTags, n=nrow(y)) # takes about four minutes

# show number of DE genes
sigsTreat_5 <- c()
for (i in seq_along(topTreat)) {
  sigsTreat_5 <- c(sigsTreat_5,sum(topTreat[[i]]$table$FDR <= 0.05))
}

sigsTreat_5 # number of DE genes per contrast
sum(sigsTreat_5)  # total number of DE genes
```

# GSEA

Get significant DE genes on 5% FDR after edgeR + TREAT

```{r}
sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast5_edgeR_DGE.txt")
cat(rownames(topTreat[[5]]$table[topTreat[[5]]$table$FDR <= 0.05,]), "\n")
sink()

sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast6_edgeR_DGE.txt")
cat(rownames(topTreat[[6]]$table[topTreat[[6]]$table$FDR <= 0.05,]), "\n")
sink()

sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast7_edgeR_DGE.txt")
cat(rownames(topTreat[[7]]$table[topTreat[[7]]$table$FDR <= 0.05,]), "\n")
sink()
# When supplying this .txt file to the GSEA engine of MSigDB, there is strong evidence for enrichment in the biological processes related to synapses, neuron projection, which are very interesting in the context of the Tasic et al. paper.
```

# Save results

```{r}
saveRDS(topTreat, file="./Data/Tasic_caseStudy_edgeR.Rds") # save to access later in comparison script
```


