---
title: "getData"
output: html_document
---

Select the cells of interest for the single cell case study based on the Tasic paper.

```{r}
metaData <- openxlsx::read.xlsx("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/Supplementary_Table_10_Full_Metadata.xlsx")

access <- read.csv2("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/GSE115746_accession_table.csv",sep = "\t")
```

```{r}
select_3 <- metaData[c(which(metaData$brain_region == "VISp" & metaData$cluster == "L5 IT VISp Hsd11b1 Endou" & metaData$eye_condition=="Normal"), which(metaData$brain_region == "ALM" & metaData$cluster == "L5 IT ALM Tmem163 Dmrtb1" & metaData$eye_condition=="Normal")),"sample_name"]

select_4 <- metaData[which(metaData$subclass =="L5 IT"  & metaData$eye_condition =="Normal" &  metaData$cluster != "L5 IT ALM Tmem163 Dmrtb1" &  metaData$cluster != "L5 IT VISp Hsd11b1 Endou"),"sample_name"]

SRR_Select_3 <- as.character(access[which(access$sample_name %in% select_3),"SRA_Run"])
SRR_Select_4 <- as.character(access[which(access$sample_name %in% select_4),"SRA_Run"])

cat(SRR_Select_3, file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/SRR_Select_3.txt",sep = "\n")
cat(SRR_Select_4, file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/SRR_Select_4.txt",sep = "\n")
```

---
Salmon quantification
Next, we retrieve the required cells from GSE and quantify the the data using salmon (add code for bash later).
---

Next, we import the salmon quantification using the txImport package.
This is done in several parts (i.e. not all cells at once) to allow for
running this script on a laptop.

```{r}
dir <- "/Volumes/ExternalDrive/DTU_project/Data/Tasic/salmon_Tasic_3"
samples_3 <- read.table(file.path(dir, "SRR_select_3.txt"), header = FALSE)
files_3 <- file.path(dir, samples_3$V1, "quant.sf")
names(files_3) <- paste0("sample", 1:length(samples_3$V1))
all(file.exists(files_3))
```

```{r, message=FALSE}
library(AnnotationHub)
## Load the annotation resource.
ah <- AnnotationHub()

## Query for all available EnsDb databases
all <- query(ah, "EnsDb")

library(ensembldb)

ahEdb <- all[["AH75036"]] #for Mus musculus

txs <- transcripts(ahEdb)
```

```{r}
tx2gene <- as.data.frame(matrix(data = NA, nrow = length(txs), ncol = 2))
colnames(tx2gene) <- c("TXNAME","GENEID")
tx2gene$TXNAME <- txs$tx_id
tx2gene$GENEID <- txs$gene_id
```

```{r}
rm(list=setdiff(ls(), c("tx2gene","files_3")))
gc()
```

```{r}
library(tximport)
quantsf_3 <- tximport(files_3, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE, txOut = TRUE)
quantsf_3_counts <- quantsf_3$counts
colnames(quantsf_3_counts) <- unname(substr(files_3, 68, 77))
```

```{r}
quantsf_3_gene <- summarizeToGene(quantsf_3, tx2gene, ignoreTxVersion = TRUE)
quantsf_3_gene_counts <- quantsf_3_gene$counts
colnames(quantsf_3_gene_counts) <- colnames(quantsf_3_counts)

rm(quantsf_3,quantsf_3_gene)
gc()
```

----

```{r}
dir <- "/Users/jg/Desktop/Phd/DTU_project/single_cell/Tasic/salmon_Tasic_4"
samples_4 <- read.table(file.path(dir, "SRR_select_4.txt"), header = FALSE)
files_4 <- file.path(dir, samples_4$V1, "quant.sf")
names(files_4) <- paste0("sample", 1:length(samples_4$V1))

sum(!file.exists(files_4))

files_4 <- files_4[file.exists(files_4)]
```

Import is done in several bits (i.e. not all cells at once) to allow for
running this script on a laptop.

```{r}
library(tximport)
# transcript level
quantsf_4_1 <- tximport(files_4[1:1000], type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE, txOut = TRUE)
quantsf_4_1_counts <- quantsf_4_1$counts
colnames(quantsf_4_1_counts) <- unname(substr(files_4, 68, 77))[1:1000]

# gene level
quantsf_4_1_gene <- summarizeToGene(quantsf_4_1, tx2gene, ignoreTxVersion = TRUE)
quantsf_4_1_gene_counts <- quantsf_4_1_gene$counts

rm(quantsf_4_1,quantsf_4_1_gene)
gc()

# transcript level
quantsf_4_2 <- tximport(files_4[1001:2000], type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE, txOut = TRUE)
quantsf_4_2_counts <- quantsf_4_2$counts
colnames(quantsf_4_2_counts) <- unname(substr(files_4, 68, 77))[1001:2000]

# gene level
quantsf_4_2_gene <- summarizeToGene(quantsf_4_2, tx2gene, ignoreTxVersion = TRUE)
quantsf_4_2_gene_counts <- quantsf_4_2_gene$counts

rm(quantsf_4_2,quantsf_4_2_gene)
gc()

# transcript level
quantsf_4_3 <- tximport(files_4[2001:2740], type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE, txOut = TRUE)
quantsf_4_3_counts <- quantsf_4_3$counts
colnames(quantsf_4_3_counts) <- unname(substr(files_4, 68, 77))[2001:2740]

# gene level
quantsf_4_3_gene <- summarizeToGene(quantsf_4_3, tx2gene, ignoreTxVersion = TRUE)
quantsf_4_3_gene_counts <- quantsf_4_3_gene$counts

rm(quantsf_4_3,quantsf_4_3_gene)
gc()
```

Combine 

```{r}
quantsf_counts <- cbind(quantsf_3_counts,quantsf_4_1_counts,quantsf_4_2_counts,quantsf_4_3_counts)
rm(quantsf_3_counts,quantsf_4_1_counts,quantsf_4_2_counts,quantsf_4_3_counts)
gc()

quantsf_counts_gene <- cbind(quantsf_3_gene_counts,quantsf_4_1_gene_counts,quantsf_4_2_gene_counts,quantsf_4_3_gene_counts)
rm(quantsf_3_gene_counts,quantsf_4_1_gene_counts,quantsf_4_2_gene_counts,quantsf_4_3_gene_counts)
gc()
```

```{r}
dim(quantsf_counts)
sum(quantsf_counts==0)/(sum(quantsf_counts==0)+sum(quantsf_counts!=0))
# 82.3% zeros on raw counts (no filter)
```

```{r}
saveRDS(quantsf_counts, file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/new_analysis/Data/Tasic_case_data/quantsf_counts_L5IT")

saveRDS(quantsf_counts_gene, file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/new_analysis/Data/Tasic_case_data/quantsf_counts_gene_L5IT")
```

