---
title: "satuRn_analysis"
author: "Jeroen Gilis"
date: "11/11/2020"
output: html_document
---

This is scripts performs a DTU analysis with satuRn on a subset of cells from the single-cell RNA-Seq dataset by Tasic et al. [Tasic paper](https://doi.org/10.1038/s41586-018-0654-5). In order to run this script, three files should be downloaded from Zenodo and placed in the data folder of this GitHub repository;

1. Tasic_caseStudy_transcript_counts.Rds (transcript-level expression matrix)
2. Tasic_metadata_1.xlsx
3. Tasic_metadata_2.csv

Note that the metadata files are identical to the one supplemented by Tasic et al., Supplementary_Table_10_Full_Metadata.xlsx and GSE115746_accession_table.csv, respectively. Also note that the raw data is available through [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746).

All figures generated in this script are available from our GitHub repository under ./Results/CaseStudy/


```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FALSE}
# for data import
library(AnnotationHub)
library(ensembldb)
library(openxlsx)

# for analysis
library(satuRn)
library(edgeR)
#library(tidyverse)
library(SummarizedExperiment)

# for visualization
library(ggplot2)
```

# Data import

## Import transcript information

```{r, message=FALSE}
ah <- AnnotationHub() # Load the annotation resource.
all <- query(ah, "EnsDb") # Query for all available EnsDb databases
ahEdb <- all[["AH75036"]] # for Mus musculus
txs <- transcripts(ahEdb)
tx2gene <- as.data.frame(matrix(data = NA, nrow = length(txs), ncol = 2))
colnames(tx2gene) <- c("TXNAME","GENEID")
tx2gene$TXNAME <- txs$tx_id
tx2gene$GENEID <- txs$gene_id
```

## Import counts

```{r}
Tasic_counts <- readRDS("./Data/Tasic_caseStudy_transcript_counts.Rds")
# Already remove genes with zero counts in all cells
Tasic_counts <- Tasic_counts[rowSums(Tasic_counts) != 0,]
dim(Tasic_counts)
```

## Import metadata

```{r}
access <- read.csv2("./Data/Tasic_metadata_2.csv",sep = "\t")
access <- access[match(colnames(Tasic_counts),access$SRA_Run),]

metaData <- openxlsx::read.xlsx("./Data/Tasic_metadata_1.xlsx")
metaData <- metaData[match(access$sample_name,metaData$sample_name),]

colnames(Tasic_counts) <- metaData$sample_name
metaData <- metaData[metaData$core_intermediate_call == "Core",] # only retain cell that were unambiguously assigned to a certain cell type (cluster)
Tasic_counts <- Tasic_counts[,metaData$sample_name]
metaData <- metaData[,c("sample_name", "brain_region","cluster")]
```

# Data wrangling

```{r}
# Remove transcripts that are the only isoform expressed of a certain gene
txInfo <- tx2gene
colnames(txInfo) <- c('isoform_id','gene_id')
rownames(txInfo) <- NULL

rownames(Tasic_counts) <- sub("\\..*", "", rownames(Tasic_counts))
txInfo <- txInfo[txInfo$isoform_id %in% rownames(Tasic_counts),]
txInfo <- subset(txInfo,duplicated(gene_id) | duplicated(gene_id, fromLast=TRUE))

Tasic_counts <- Tasic_counts[which(rownames(Tasic_counts) %in% txInfo$isoform_id),]
```

# Set up the experimental design

```{r}
metaData$cluster <- gsub(" ", "_", metaData$cluster)
metaData$group <- paste(metaData$brain_region,metaData$cluster,sep=".")

# Remove groups (cell types) with less than 30 cells --> removes 4 groups
remove <- names(table(interaction(metaData$brain_region,metaData$cluster))[table(interaction(metaData$brain_region,metaData$cluster)) < 30])

metaData <- metaData[-which(metaData$group %in% remove),]
group <- as.factor(metaData$group)
 
design <- model.matrix(~ 0 + group) # Factorial design
colnames(design) <- levels(group)

Tasic_counts <- Tasic_counts[,which(colnames(Tasic_counts) %in% metaData$sample_name)]
```

# Filtering

```{r}
filter_all_edgeR_stringent <- filterByExpr(Tasic_counts,
             design = NULL,
             group = metaData$brain_region,
             lib.size = NULL,
             min.count = 10,
             min.total.count = 0,
             large.n = 0,
             min.prop = 0.7)

table(filter_all_edgeR_stringent)
Tasic_counts <- Tasic_counts[filter_all_edgeR_stringent,]
```

```{r}
# Update txInfo according to the filtering procedure
txInfo <- txInfo[which(txInfo$isoform_id %in% rownames(Tasic_counts)),]

# remove transcripts that are the only isoform expressed of a certain gene (after filtering)
txInfo <- subset(txInfo,duplicated(gene_id) | duplicated(gene_id, fromLast=TRUE))

Tasic_counts <- Tasic_counts[which(rownames(Tasic_counts) %in% txInfo$isoform_id),]

# satuRn require the transcripts in the rowData and the transcripts in the count matrix to be in the same order. If not, the resulting models will be matched to the wrong rowData
txInfo <- txInfo[match(rownames(Tasic_counts),txInfo$isoform_id),]

rm(list = setdiff(ls(), c("Tasic_counts", "txInfo", "design", "metaData", "tx2gene","group")))
invisible(gc())
```

# Analysis

Make a summarized experiment object

```{r, message=F}
sumExp <- SummarizedExperiment::SummarizedExperiment(assays=list(counts=Tasic_counts), colData = metaData, rowData = txInfo)
#metadata(sumExp)$formula <- ~ 0 + group  # specify design, internally the fit function will then use design <- model.matrix(sumExp@metadata$formula, colData(sumExp))
```

## Fit quasi-binomial GLMs with satuRn

```{r}
Sys.time()
sumExp <- satuRn::fitDTU(object = sumExp,
                         formula = ~ 0 + group,
                         parallel = FALSE,
                         BPPARAM = BiocParallel::bpparam(),
                         verbose = TRUE
                         )
Sys.time()
```

## Set up contrast matrix

```{r}
# We here manually construct the contrast of interest as defined by Tasic et al.
design <- model.matrix(~0+group)
colnames(design) <- levels(group)

L <- matrix(0, ncol = 8, nrow = ncol(design))
rownames(L) <- colnames(design)
colnames(L) <- c("C1", "C2", "C3", "C4", "C5", "C6", "C7", "C8")

L[c("VISp.L5_IT_VISp_Batf3", "ALM.L5_IT_ALM_Cpa6_Gpr88"), 1] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col27a1", "ALM.L5_IT_ALM_Cbln4_Fezf2"), 2] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col6a1_Fezf2", "ALM.L5_IT_ALM_Cpa6_Gpr88"), 3] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Col6a1_Fezf2", "ALM.L5_IT_ALM_Gkn1_Pcdh19"), 4] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Lypd1_Gpr88"), 5] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Tnc"), 6] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Hsd11b1_Endou", "ALM.L5_IT_ALM_Tmem163_Dmrtb1"), 7] <- c(1,-1)
L[c("VISp.L5_IT_VISp_Whrn_Tox2", "ALM.L5_IT_ALM_Tmem163_Arhgap25"), 8] <- c(1,-1)
```

## Test contrasts with satuRn

```{r}
sumExp <- satuRn::testDTU(object = sumExp, contrasts = L, plot=T, sort = F)
```

```{r}
# now the rowdata have 8 additional slots, corresponding to the 8 different requested contrasts. This is a simple dataframe with results for each transcripts in the selected contrast.
rowData(sumExp)[["fitDTUResult_C6"]]
```

# Visualize DTU with satuRn

To create panels A, B and C of Figure 6 of our paper.

```{r}
group1 <- rownames(colData(sumExp))[colData(sumExp)$group == "VISp.L5_IT_VISp_Hsd11b1_Endou"]
group2 <- rownames(colData(sumExp))[colData(sumExp)$group == "ALM.L5_IT_ALM_Tnc"]

plots <- satuRn::plotDTU(object = sumExp, contrast = "C6", groups = list(group1,group2), coefficients = list(c(0,0,0,0,0,0,0,0,0,0,0,0,1,0),c(0,0,0,0,0,0,0,0,1,0,0,0,0,0)), summaryStat = "model", transcripts = c("ENSMUST00000081554","ENSMUST00000195963","ENSMUST00000132062"),genes = NULL,top.n = 6)

# To have same layout as in our paper
for (i in seq_along(plots)) {
    titles <- c("Figure6A","Figure6B","Figure6C")
    current_plot <- plots[[i]]  + 
    scale_fill_manual(labels = c("VISp","ALM"), values=c("royalblue4", "firebrick")) +
    scale_x_discrete(labels= c("Hsd11b1_Endou","Tnc")) +
    theme(axis.text.x = element_text(angle = 0, vjust = 1, hjust = 0.5, size = 9)) + 
    theme(strip.text = element_text(size = 9, face = "bold"))
  
    png(paste0("./Results/CaseStudy/",titles[i],".png"),
        width     = 5,
        height    = 5,
        units     = "in",
        res       = 200,
        pointsize = 4)
    print(current_plot)
    dev.off()
}
```

# GSEA

Save DTU transcripts for contrasts 5, 6 and 7 to perform a DTU analysis (import to MSigDB)

```{r}
DTU_tx <-  rownames(rowData(sumExp)[[5+3]][rowData(sumExp)[[5+3]]$empirical_FDR <= 0.05 & !is.na(rowData(sumExp)[[5+3]]$estimates),])
DTU_gene <- tx2gene[which(tx2gene$TXNAME %in% DTU_tx), "GENEID"]
sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast5_DTU_satuRn.txt")
cat(DTU_gene, "\n")
sink()

DTU_tx <-  rownames(rowData(sumExp)[[6+3]][rowData(sumExp)[[6+3]]$empirical_FDR <= 0.05 & !is.na(rowData(sumExp)[[6+3]]$estimates),])
DTU_gene <- tx2gene[which(tx2gene$TXNAME %in% DTU_tx), "GENEID"]
sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast6_DTU_satuRn.txt")
cat(DTU_gene, "\n")
sink()

DTU_tx <-  rownames(rowData(sumExp)[[7+3]][rowData(sumExp)[[7+3]]$empirical_FDR <= 0.05 & !is.na(rowData(sumExp)[[7+3]]$estimates),])
DTU_gene <- tx2gene[which(tx2gene$TXNAME %in% DTU_tx), "GENEID"]
sink(file = "./Results/CaseStudy/GSEA/Tasic_L5IT_contrast7_DTU_satuRn.txt")
cat(DTU_gene, "\n")
sink()
```

# Save final object

Save resulting object for later comparison of results with limma diffsplice and DoubleExpSeq.

```{r}
saveRDS(sumExp, file = "./Data/Tasic_caseStudy_satuRn.Rds")
```




