---
title: "Love_ECC_lengthBias"
output: html_document
---

Length bias analysis for the count dataset

```{r}
library(grid)
library(gridExtra)
library(ggplot2)
library(rtracklayer)
library(IsoformSwitchAnalyzeR)
```

```{r}
load(file='01_love_et_al_benchmark_data_ECC_counts.Rdata')
#load(file='01_love_et_al_benchmark_data_ECC_tpm.Rdata')

### Combine
names(loveBenchmarkDataLenient)   <- paste0(names(loveBenchmarkDataLenient)  , '_filterLenient')
names(loveBenchmarkDataStringent) <- paste0(names(loveBenchmarkDataStringent), '_filterStringent')
metaInfo <- c(
    loveBenchmarkDataLenient,
    loveBenchmarkDataStringent
)

load('02_love_et_al_benchmark_DTU_analysis_ECC_count.Rdata')
#load('02_love_et_al_benchmark_DTU_analysis_ECC_tpm.Rdata')
    
LoveBenchmark_count <- c(
   loveDtuBenchmark_DEXSeq,
   loveDtuBenchmark_DoubleExpSeq,
   loveDtuBenchmark_DRIMSeq,
   loveDtuBenchmark_edgeRdiffsplice,
   loveDtuBenchmark_limmaDiffsplice,
   loveDtuBenchmark_NBSplice,
   loveDtuBenchmark_quasiBinomial)

LoveBenchmark_count  <- LoveBenchmark_count[which(
        sapply(LoveBenchmark_count, function(x) ! is.null(x$dtuAnalysis))
)]

# remove spurious objects
rm(loveBenchmarkDataLenient,
   loveBenchmarkDataStringent,
   loveDtuBenchmark_DEXSeq,
   loveDtuBenchmark_DRIMSeq,
   loveDtuBenchmark_DoubleExpSeq,
   loveDtuBenchmark_edgeRdiffsplice,
   loveDtuBenchmark_limmaDiffsplice,
   loveDtuBenchmark_NBSplice,
   loveDtuBenchmark_quasiBinomial)
gc()
```

Get isoform length information (human)

```{r}
# gencode annotation version 28 was used by Love et al. for generating the simulation study
gtf_h <- rtracklayer::import("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Benchmarks/Bulk_simulations/Love_dataset_ECC/gencode.v28.annotation.gtf")
gtf_h <- gtf_h[which(gtf_h$type == 'exon'),]

isoformLengths <- sapply(
    X = split(
        x = gtf_h@ranges@width,
        f = gtf_h$transcript_id
    ),
    FUN = sum
)

txs_len <- isoformLengths
names(txs_len) <- sub("\\..*", "", names(isoformLengths))
```

Compute transcript length bias for each method in each dataset

```{r}
resList <- list()

# loop over datasets
for (current_dataset in names(metaInfo)) {
  
  # select current dataset
  LoveBenchmark_count_current <- LoveBenchmark_count[grepl(current_dataset,names(LoveBenchmark_count))]
  metaInfo_current <- metaInfo[grepl(current_dataset,names(metaInfo))]
  
  # select transcripts in current dataset
  txs_len_current <- txs_len[which(names(txs_len) %in% sub("\\..*", "", metaInfo_current[[1]]$metaInfo$TXNAME))]
  txs_len_current <- txs_len_current[unique(names(txs_len_current))]
  
  # define "long" trascript as longer than median
  long <- ifelse(txs_len_current>median(txs_len_current),"yes","no")
  
  # For DoubleExpSeq and NBSplice, FDRs must still be calculated
  LoveBenchmark_count_current[[2]]$dtuAnalysis$FDR <- p.adjust(LoveBenchmark_count_current[[2]]$dtuAnalysis$p_value, method = "BH")
  LoveBenchmark_count_current[[6]]$dtuAnalysis$FDR <- p.adjust(LoveBenchmark_count_current[[6]]$dtuAnalysis$p_value, method = "BH")

  # get results for each of the methods
  ### start getting result
  stats <- c()
  pvalues <- c()
  
  for (i in 1:length(LoveBenchmark_count_current)) {

      method_result <- ifelse(LoveBenchmark_count_current[[i]]$dtuAnalysis$FDR <= 0.05, "yes", "no")
      method_test <- fisher.test(table(long,method_result))
      stats <- c(stats,method_test$estimate)
      pvalues <- c(pvalues,method_test$p.value)

  }
  
  # wrangle results
  methods <- c("DEXSeq","DoubleExpSeq","DRIMSeq","edgeR_diffsplice","limma_diffsplice","NBSplice","qbDTU")
  
  current_result <- data.frame(cbind(methods,stats,pvalues))
  colnames(current_result) <- c("method","stats","pvalues")

  current_result$method <- as.factor(current_result$method)
  current_result$stats <- as.numeric(current_result$stats)
  current_result$pvalues <- as.numeric(current_result$pvalues)
  rownames(current_result) <- 1:nrow(current_result)
  
  # save result in list
  resList[[current_dataset]] <- current_result
}
```

Plot performances

```{r}
analyses <- c("3_rep_._filterLenient","6_rep_._filterLenient","10_rep_._filterLenient","3_rep_._filterStringent","6_rep_._filterStringent","10_rep_._filterStringent")

plotList <- list()

for (analysis in analyses) {
  
  current_analysis <- resList[grepl(analysis,names(resList))]

  gg <- ggplot(data = current_analysis[[1]],mapping=aes(x=stats,y=method,color=ifelse(pvalues<0.05, "red", "black"))) +
      scale_color_identity() +
      geom_point() +
      geom_point(data=current_analysis[[2]], mapping=aes(x=stats,y=method)) +
      geom_point(data=current_analysis[[3]], mapping =aes(x=stats,y=method)) +
      theme_bw() +
      ggtitle(analysis)
  
  plotList[[analysis]] <- gg
}
# png("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Results/Love_benchmark/plot_count_lengthBias.png",
#     width     = 8,
#     height    = 5,
#     units     = "in",
#     res       = 200,
#     pointsize = 4) # start export
do.call(grid.arrange, c(plotList, ncol=3))
# dev.off()
```








