---
title: "1_Chen_prepare"
author: "Jeroen Gilis"
date: "03/11/2020"
output: html_document
---

**In order to run this script (1_Chen_perpare.Rmd), the corresponding expression count file Chen_counts.Rds (or, alternatively, Chen_scaledTPM.Rds) be downloaded from Zenodo. In addition, the metaData Chen_metadata.csv should also be downloaded from Zenodo. Both files must then be copied into the Data folder of this Github page**.

**If one does not want to run this script, its output can also be downloaded from Zenodo: Chen_benchmark_datasets_count.Rdata (or, alternatively, Chen_benchmark_datasets_scaledTPM.Rdata) **

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FALSE}
library(edgeR)
library(DRIMSeq)
library(AnnotationHub)
library(ensembldb)
library(doMC)
```

# Load data and metadata

```{r}
Chen_counts <- readRDS(file = "./Data/Chen_counts.Rds")
#Chen_counts <- readRDS(file ="./Data/Chen_scaledTPM.Rds")

## remove transcripts with zero count for all cells
Chen_counts <- Chen_counts[which(rowSums(Chen_counts)!=0),]

metaData <- read.csv("./Data/Chen_metadata.csv")
rownames(metaData) <- metaData$Run
metaData <- metaData[colnames(Chen_counts),]
```

# Load genome annotation

```{r, message=FALSE}
## Load the annotation resource.
ah <- AnnotationHub()

## Query for all available EnsDb databases
all <- query(ah, "EnsDb")
ahEdb <- all[["AH75036"]] #for Mus musculus

txs <- transcripts(ahEdb)

### Extract gene info
tx2gene <- as.data.frame(matrix(data = NA, nrow = length(txs), ncol = 2))
colnames(tx2gene) <- c("TXNAME","GENEID")
tx2gene$TXNAME <- txs$tx_id
tx2gene$GENEID <- txs$gene_id

txInfo <- tx2gene
colnames(txInfo) <- c('isoform_id','gene_id')
rownames(txInfo) <- NULL

txInfo <- txInfo[which(
    ! grepl('^ERCC-', txInfo$isoform_id )
),]

txInfo <- txInfo[which(txInfo$isoform_id %in% sub("\\..*", "", rownames(Chen_counts))),]

txInfo$isoform_id <- rownames(Chen_counts)[match(txInfo$isoform_id,sub("\\..*", "", rownames(Chen_counts)))]
```

# Subset Chen data

To generate a benchmark dataset, we need to know the ground truth, i.e. which transcripts are differentially used between groups of samples. To this end, we first generate mock datasets where no DTU is expected, after which we introduce DTU between groups by the transcript-abundance-swapping strategy discussed by Van den Berge et al. [stageR paper](https://doi.org/10.1186/s13059-017-1277-0).

In order to obtain a mock dataset, we need to construct a dataset where (ideally) no DTU is expected. We try to achieve this by subsampling a homogenous set of samples from the Chen dataset, based on following criteria:

1. metaData$Cell_type == "Epi"
2. metaData$sex == "female"

```{r}
## Select a homogeneous set of cells
metaData <- metaData[which(metaData$Cell_type == "Epi" & metaData$sex == "female"),]
Chen_counts <- Chen_counts[,as.character(metaData$Run)]
```

# Setup generation of benchmark data

Set hyperparameters of the benchmark datasets (i.e. sample size, number of repeats, fraction of DTU genes and parameters for parallel processing).

```{r}
### Set parameters
samplesPrCondition   <- c(20,50)
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

### Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'

    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

### list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
```

# Generate gtex benchmark data

```{r}
source(file="./Performance_benchmarks/getBenchmark_data.R")
```

```{r}
ChenBenchmarkLenient <- getBenchmark_data(countData=Chen_counts, 
                                          metaData=txInfo,
                                          filter="edgeR",
                                          edgeR_filter_spec = list(min.count = 1, 
                                                                    min.total.count = 0, 
                                                                    large.n = 0, 
                                                                    min.prop = 0.5),
                                          nrRepList=nrRepList, 
                                          fracGenesAffected=0.15)
names(ChenBenchmarkLenient) <- paste0(names(nrRepList),"_filterLenient")

ChenBenchmarkStringent <- getBenchmark_data(countData=Chen_counts,
                                                metaData=txInfo,
                                                filter="DRIMSeq", 
                                                nrRepList=nrRepList, 
                                                fracGenesAffected=0.15)
names(ChenBenchmarkStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save Chen benchmark data

```{r}
#save(ChenBenchmarkLenient, ChenBenchmarkStringent, file="./Data/Chen_benchmark_datasets_count.Rdata")
save(ChenBenchmarkLenient, ChenBenchmarkStringent, file="./Data/Chen_benchmark_datasets_scaledTPM.Rdata")
```



