---
title: "1_Tasic_prepare"
author: "Jeroen Gilis"
date: "05/11/2020"
output: html_document
---

**In order to run this script (1_Tasic_perpare.Rmd), the corresponding salmon quantification files Tasic_counts.Rds should either be obtained by running 0_Tasic_getData.Rmd or downloaded from Zenodo. In addition, two metaData files Tasic_metadata_1.xlsx, Tasic_metadata_2.csv should also be downloaded from Zenodo. All three files must then be copied into the Data folder of this Github page**. Note that the metadata files are identical to the one supplemented in the paper by Tasic et al. [Tasic paper](https://doi.org/10.1038/s41586-018-0654-5), Supplementary_Table_10_Full_Metadata.xlsx and GSE115746_accession_table.csv, respectively.

1. Tasic_counts.Rds
2. Tasic_metadata.xlsx

**If one does not want to run this script, its output can also be downloaded from Zenodo: 02_Tasic_benchmark_datasets_count.Rdata**

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FASLE}
library(edgeR)
library(DRIMSeq)
library(openxlsx)
library(AnnotationHub)
library(ensembldb)
library(doMC)
```

# Load data and metadata

```{r}
Tasic_counts <- readRDS(file = "./Data/Tasic_counts.Rds")
# Tasic_counts <- readRDS(file = "./Data/Tasic_scaledTPM.Rds")

metaData_1 <- openxlsx::read.xlsx("./Data/Tasic_metadata_1.xlsx")
metaData_2 <- read.csv2("./Data/Tasic_metadata_2.csv",sep = "\t")
```

# Load genome annotation

```{r, message=FALSE}
## Load the annotation resource.
ah <- AnnotationHub()

## Query for all available EnsDb databases
all <- query(ah, "EnsDb")

ahEdb <- all[["AH75036"]] #for Mus musculus

txs <- transcripts(ahEdb)

tx2gene <- as.data.frame(matrix(data = NA, nrow = length(txs), ncol = 2))
colnames(tx2gene) <- c("TXNAME","GENEID")
tx2gene$TXNAME <- txs$tx_id
tx2gene$GENEID <- txs$gene_id

### Extract gene info
txInfo <- tx2gene
colnames(txInfo) <- c('isoform_id','gene_id')
rownames(txInfo) <- NULL

txInfo <- txInfo[which(
    ! grepl('^ERCC-', txInfo$isoform_id )
),]
```

# Subset Tasic data

To generate a benchmark dataset, we need to know the ground truth, i.e. which transcripts are differentially used between groups of samples. To this end, we first generate mock datasets where no DTU is expected, after which we introduce DTU between groups by the transcript-abundance-swapping strategy discussed by Van den Berge et al. [stageR paper](https://doi.org/10.1186/s13059-017-1277-0).

In order to obtain a mock dataset, we need to construct a dataset where (ideally) no DTU is expected. Therefore, we only quantified cells that corresponded to the following criteria:

1. metaData_1$subclass == "Lamp5"
2. metaData_1$brain_region == "ALM"
3. metaData_1$eye_condition=="Normal"

In additon, we now further subsample the cells to meet the criterium

4. metaData_1$cluster == "Lamp5 Lsp1"

```{r}
## remove transcripts with zero count for all cells
Tasic_counts <- Tasic_counts[which(rowSums(Tasic_counts)!=0),]

metaData_1 <- metaData_1[metaData_1$sample_name %in% metaData_2[which(metaData_2$SRA_Run %in% colnames(Tasic_counts)),"sample_name"],]

metaData_1 <- metaData_1[metaData_1$cluster == "Lamp5 Lsp1",]
metaData_2 <- metaData_2[which(metaData_2$sample_name %in% metaData_1$sample_name),]
Tasic_counts <- Tasic_counts[,colnames(Tasic_counts)%in%metaData_2$SRA_Run]

## make rownames count frame compatible with those of txInfo frame
rownames(Tasic_counts) <- sub("\\..*", "", rownames(Tasic_counts))
```

# Setup generation of benchmark data

Set hyperparameters of the benchmark datasets (i.e. sample size, number of repeats, fraction of DTU genes and parameters for parallel processing).

```{r}
### Set parameters
samplesPrCondition   <- c(20,75,200)
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

### Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'
    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

### list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
```

# Generate Tasic benchmark data

```{r}
source(file="./Performance_benchmarks/getBenchmark_data.R")
```

```{r}
TasicBenchmarkLenient <- getBenchmark_data(countData=Tasic_counts, 
                                          metaData=txInfo,
                                          filter="edgeR",
                                          edgeR_filter_spec = list(min.count = 1, 
                                                                    min.total.count = 0, 
                                                                    large.n = 0, 
                                                                    min.prop = 0.5),
                                          nrRepList=nrRepList, 
                                          fracGenesAffected=0.15)
names(TasicBenchmarkLenient) <- paste0(names(nrRepList),"_filterLenient")

TasicBenchmarkStringent <- getBenchmark_data(countData=Tasic_counts,
                                                metaData=txInfo,
                                                filter="DRIMSeq", 
                                                nrRepList=nrRepList, 
                                                fracGenesAffected=0.15)
names(TasicBenchmarkStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save Tasic benchmark data

```{r}
save(TasicBenchmarkLenient, TasicBenchmarkStringent, file="./Data/02_Tasic_benchmark_datasets_count.Rdata")
#save(TasicBenchmarkLenient, TasicBenchmarkStringent, file="./Data/02_Tasicbenchmark_datasets_scaledTPM.Rdata")
```


