---
title: "Tasic_prepare"
output: html_document
---

```{r,message=FALSE}
library(edgeR)
library(DRIMSeq)
```

```{r}
# quantsf_counts <- readRDS(file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/quantsf_scaled_TPM")

quantsf_counts <- readRDS(file = "/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/quantsf_counts")

## remove transcripts with zero count for all cells
quantsf_counts <- quantsf_counts[which(rowSums(quantsf_counts)!=0),]

metaData <- openxlsx::read.xlsx("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/Supplementary_Table_10_Full_Metadata.xlsx")

access <- read.csv2("/Users/jg/Desktop/PhD/DTU_project/Github/dtuPaper/Data/Tasic_data/GSE115746_accession_table.csv",sep = "\t")

##  make sure this is unnecessary!!!
dir <- "/Volumes/ExternalDrive/DTU_project/Data/Tasic/salmon_Tasic_1"
samples <- read.table(file.path(dir, "SRR_select.txt"), header = FALSE)
files <- file.path(dir, samples$V1, "quant.sf")
names(files) <- paste0("sample", 1:length(samples$V1))
files <- files[which(file.exists(files))]
selected <- unname(substring(files, 62,71))

colnames(quantsf_counts) <- selected

metaData <- metaData[metaData$sample_name %in% access[which(access$SRA_Run %in% selected),"sample_name"],]

metaData <- metaData[metaData$cluster == "Lamp5 Lsp1",]
access <- access[which(access$sample_name %in% metaData$sample_name),]
quantsf_counts <- quantsf_counts[,colnames(quantsf_counts)%in%access$SRA_Run]

rm(access,samples,dir,files,selected,metaData)
invisible(gc())
```


```{r, message=FALSE}
library(AnnotationHub)
## Load the annotation resource.
ah <- AnnotationHub()

## Query for all available EnsDb databases
all <- query(ah, "EnsDb")

library(ensembldb)

ahEdb <- all[["AH75036"]] #for Mus musculus

txs <- transcripts(ahEdb)

tx2gene <- as.data.frame(matrix(data = NA, nrow = length(txs), ncol = 2))
colnames(tx2gene) <- c("TXNAME","GENEID")
tx2gene$TXNAME <- txs$tx_id
tx2gene$GENEID <- txs$gene_id
```

```{r}
## quantsf_counts holds my counts, as obtained above
## tx2gene holds the gene info, as obtained in the Tasic_txImport.Rmd script

### Extract gene info
txInfo <- tx2gene
colnames(txInfo) <- c('isoform_id','gene_id')
rownames(txInfo) <- NULL

txInfo <- txInfo[which(
    ! grepl('^ERCC-', txInfo$isoform_id )
),]

## make rownames count frame compatible with those of txInfo frame
rownames(quantsf_counts) <- sub("\\..*", "", rownames(quantsf_counts))

### Set parameters
samplesPrCondition   <- c(20,75,200)
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

### Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'

    library(doMC)
    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

### list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
### Sanity check
length(nrRepList) == length(samplesPrCondition) * nrRepsMade
table(sapply(nrRepList, length))
```

Generate lenient benchmark

```{r}
TasicBenchmarkLenient <- lapply(c(1:length(nrRepList)), function(x) {
      
    ### Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]

    if(TRUE) {
        localSubset <- sample(
            colnames(quantsf_counts),
            localSampleSize * 2
        )

        localDesign <- data.frame(
            sample_id = localSubset,
            condition = c(
                rep('a',floor  ( length(localSampleSize) )),
                rep('b',ceiling( length(localSampleSize) ))
            ),
            stringsAsFactors = FALSE
        )
        localDesign <- localDesign[sort.list(localDesign$condition),]
    }
      
        ### Step 2: Subset to expressed features using edgeR::filterByExpr
        if(TRUE) {
            
            y <- edgeR::DGEList(counts = quantsf_counts[,localDesign$sample_id])
            design <- model.matrix(~condition, data=localDesign)
            
            filter <- edgeR::filterByExpr(y,design=design,min.prop=0.5,min.count=1,min.total.count = 0,large.n=0) 
            
            localCm <- y$counts[filter,]
            ## Get only multi-isoform genes (after filtering)
            localTx <- txInfo[which(
                txInfo$isoform_id %in% rownames(localCm)),]
            
            tmp <- table(localTx$gene_id)
            tmp <- tmp[which( tmp >= 2)]
        
            localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
            localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
        }

        ### Step 3: Extract isoforms to modify
        if(TRUE) {
            genesToModify <- sample(
                x = unique(localTx$gene_id),
                size = round(
                    length(unique(localTx$gene_id)) * fracGenesAffected
                )
            )
            
            samplesToModify <- localDesign$sample_id[which(
                localDesign$condition == 'b')
                ]
            
            cm_swapped <- localCm
            
            transcripts_toSwap_current <- c()
            transcripts_swapped_current <- c()
            transcripts_swapped_all <- c()
            transcripts_toSwap_all <- c()

            for (gene in genesToModify) {

                current <- localTx[which(localTx$gene_id==gene),]
                nSwap <- max(2,rbinom(1,nrow(current),1/3))
  
                transcripts_toSwap_current <- sample(
                    x = current$isoform_id,
                    size = nSwap)
  
                ## swap order of txs completely by putting the fist one last
                transcripts_swapped_current <- c(transcripts_toSwap_current[-1],transcripts_toSwap_current[1])
                
                ## Add to swapping queue
                transcripts_swapped_all <- c(transcripts_swapped_all,transcripts_swapped_current)
                transcripts_toSwap_all <- c(transcripts_toSwap_all,transcripts_toSwap_current)
            }
        
        ## Perform the swapping in matrix
        cm_swapped[transcripts_toSwap_all,which(colnames(cm_swapped)%in%samplesToModify)] <- cm_swapped[transcripts_swapped_all,which(colnames(cm_swapped)%in%samplesToModify)]
      
        ## Set swapping in localTx
        localTx$txSwapped <- vector(length=nrow(localTx))
        localTx[which(localTx$isoform_id %in% transcripts_swapped_all),"txSwapped"] <- TRUE 
        localTx$nrSamplesPerCondition <- localSampleSize
        }
    
        colnames(localTx) <- c('TXNAME','GENEID','txSwapped','nrSamplesPerCondition')

        ### Combine data
        dataList <- list(
            data     = cm_swapped,
            design   = localDesign,
            metaInfo = localTx
        )
        return(dataList)
    }
    
)

names(TasicBenchmarkLenient) <-  paste0(names(nrRepList),"_filterLenient")
```

```{r}
for (i in 1:length(TasicBenchmarkLenient)) {
  print(dim(TasicBenchmarkLenient[[i]]$data))
}
```

Generate Stringent benchmark

```{r}
TasicBenchmarkStringent <- lapply(c(1:length(nrRepList)), function(x) {
      
    ### Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]

    if(TRUE) {
        localSubset <- sample(
            colnames(quantsf_counts),
            localSampleSize * 2
        )

        localDesign <- data.frame(
            sample_id = localSubset,
            condition = c(
                rep('a',floor  ( length(localSampleSize) )),
                rep('b',ceiling( length(localSampleSize) ))
            ),
            stringsAsFactors = FALSE
        )
        localDesign <- localDesign[sort.list(localDesign$condition),]
    }
      
        ### Step 2: Subset to expressed features using DRIMSeq::dmFilter
        if(TRUE) {
          
          localCm <-  as.data.frame(quantsf_counts[,localDesign$sample_id])
          
          geneForEachTx <- txInfo[match(rownames(localCm),txInfo[,"isoform_id"]),"gene_id"]

          localCm$gene_id <- geneForEachTx
          localCm$feature_id <- row.names(localCm)

          d <- DRIMSeq::dmDSdata(counts = localCm, samples = localDesign)

          d_filter <- dmFilter(d,
                min_samps_feature_expr=localSampleSize/2, min_feature_expr=10, min_samps_feature_prop=localSampleSize/2, min_feature_prop=0.1,
                min_samps_gene_expr=localSampleSize, min_gene_expr=10)

          localCm <- quantsf_counts[counts(d_filter)$feature_id,localDesign$sample_id]

          ## Get only multi-isoform genes (after filtering)
          localTx <- txInfo[which(
              txInfo$isoform_id %in% rownames(localCm)),]
          
          tmp <- table(localTx$gene_id)
          tmp <- tmp[which( tmp >= 2)]
        
          localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
          localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
        }

        ### Step 3: Extract isoforms to modify
        if(TRUE) {
            genesToModify <- sample(
                x = unique(localTx$gene_id),
                size = round(
                    length(unique(localTx$gene_id)) * fracGenesAffected
                )
            )
            
            samplesToModify <- localDesign$sample_id[which(
                localDesign$condition == 'b')
                ]
            
            cm_swapped <- localCm
            
            transcripts_toSwap_current <- c()
            transcripts_swapped_current <- c()
            transcripts_swapped_all <- c()
            transcripts_toSwap_all <- c()

            for (gene in genesToModify) {

                current <- localTx[which(localTx$gene_id==gene),]
                nSwap <- max(2,rbinom(1,nrow(current),1/3))
  
                transcripts_toSwap_current <- sample(
                    x = current$isoform_id,
                    size = nSwap)
  
                ## swap order of txs completely by putting the fist one last
                transcripts_swapped_current <- c(transcripts_toSwap_current[-1],transcripts_toSwap_current[1])
                
                ## Add to swapping queue
                transcripts_swapped_all <- c(transcripts_swapped_all,transcripts_swapped_current)
                transcripts_toSwap_all <- c(transcripts_toSwap_all,transcripts_toSwap_current)
            }
        
        ## Perform the swapping in matrix
        cm_swapped[transcripts_toSwap_all,which(colnames(cm_swapped)%in%samplesToModify)] <- cm_swapped[transcripts_swapped_all,which(colnames(cm_swapped)%in%samplesToModify)]
      
        ## Set swapping in localTx
        localTx$txSwapped <- vector(length=nrow(localTx))
        localTx[which(localTx$isoform_id %in% transcripts_swapped_all),"txSwapped"] <- TRUE 
        localTx$nrSamplesPerCondition <- localSampleSize
        }
    
        colnames(localTx) <- c('TXNAME','GENEID','txSwapped','nrSamplesPerCondition')

        ### Combine data
        dataList <- list(
            data     = cm_swapped,
            design   = localDesign,
            metaInfo = localTx
        )

        return(dataList)
    }
    
)

names(TasicBenchmarkStringent) <- paste0(names(nrRepList),"_filterStringent")
```

```{r}
save(TasicBenchmarkLenient, TasicBenchmarkStringent, file='01_Tasic_benchmark_data_count.Rdata')
#save(TasicBenchmarkLenient, TasicBenchmarkStringent, file='01_Tasic_benchmark_data_scaledTPM.Rdata')
```



