## satuRn: Scalable Analysis of Differential Transcript Usage for Bulk and Single-cell RNA-sequencing Applications

### Availability of data

The datasets required to reproduce all results that are displayed in this publication (including supplementary materials) are available at Zenodo. This includes both the raw data and intermediate results. Note that at the top of each analysis script it is indicated which dataset is required as input for the script; it may thus not be necessary to download all datasets from Zenodo. For easily reproducing the results, place the downloaded data in this Data folder.

INSERT ZENODO LINK

***

### Analyses & Scripts

To reproduce the results that are displayed in this publication, proceed as follows:

1. Make a local clone of this Github repository
2. Open the selected R scripts (.Rmd) in the R project "DTU_paper.Rproj" in the root of this repository.
3. Look at the top of the R scripts what input data is required. Download this data from Zenodo (see Availability of data) and place the data in the `Data`  folder of this repository.
4. Run the analyses - the results will automatically be stored in the repository, either in the "Data" folder (if it is an intermediate results) or in the Results folder (if it is an end resuls, e.g. a figure.)

This Github package contains the following folders:

- Data: For reproducing the results displayed in this publication, input data for the different analysis should be downloaded from Zenodo and placed in the Data folder of *your local clone* of this Github page.

1. Performance_benchmarks: This folder contains all the code






















