---
title: "Gtex_prepare_filter"
output: html_document
---

# Load libraries

```{r,message=FALSE,warning=FALSE}
library(data.table)
library(edgeR)
library(DRIMSeq)
```

# Download Gtex data

The raw data files for the Gtex benchmark were downloaded from:

```{r}
#https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt
#https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RSEMv1.3.0_transcript_expected_count.gct.gz
```

**In order to run this specific script (1_Gtex_perpare.Rmd), these files should be downloaded from Zenodo and copied into the raw_data folder from this Github page**, which is located in the directory /satuRn_paper/Data. However, not that the other scripts (2_Gtex_DTU.Rmd and 3_Gtex_visualize.Rmd) can be run with intermediate data files that are already present on this GitHub page.

# Load Gtex data

```{r,warning=FALSE}
err <- try(library("here", character.only = TRUE), silent = TRUE)
if (class(err) == 'try-error') {
    install.packages("here", repos = "https://cloud.r-project.org")
    library("here", character.only = TRUE)
}
wd <- here()
# Optional: change the working directory
setwd(wd)

gtexCm <- fread(paste0(wd, "/Data/raw_data/GTEx_Analysis_2017-06-05_v8_RSEMv1.3.0_transcript_expected_count.gct"), data.table = FALSE)
gtexSample <- fread(paste0(wd, "/Data/raw_data/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"), data.table = FALSE)
```

# Subset Gtex data

The Gtex dataset comes from an observational study. To use it as a benchmark dataset, we need to know  the ground truth, i.e. which transcripts are differentially used between groups of samples. To this end we first generate mock datasets where no DTU is expected and second we introduce DTU between groups by the transcript-abundance-swapping discussed by Van den Berge et al. [stageR paper](https://doi.org/10.1186/s13059-017-1277-0).

In order  to obtain a mock dataset, however, we need to construct a dataset where (ideally) no DTU is expected. We try to achieve this by subsampling a homogenous set of samples from the Gtex data object, based on following criteria:

- Only adrenal gland data 
- Only RNA Extraction from Paxgene-derived Lysate Plate Based samples
- Only samples from center B1
- Only RNASEQ samples in GTEx Analysis Freeze (SMAFRZE)

In aaddition, in the code below we also set benchmark hyperparameters (i.e. sample size, number of repeats)

```{r}
# Extract gene info
txInfo <- gtexCm[,c('transcript_id','gene_id')]
colnames(txInfo)[1] <- 'isoform_id'
rownames(gtexCm) <- txInfo$isoform_id

# Extract samples
gtexSample <- gtexSample[
    which(gtexSample$SMTSD == 'Adrenal Gland'),
]

# specifically filter on extraction kit
gtexSample <- gtexSample[
    which(gtexSample$SMNABTCHT == 'RNA Extraction from Paxgene-derived Lysate Plate Based'),
]

# specifically filter on SMAFRZE
gtexSample <- gtexSample[
    which(gtexSample$SMAFRZE == 'RNASEQ'),
]

# specifically filter on center
gtexSample <- gtexSample[
    which(gtexSample$SMCENTER == 'B1'),
]

gtexCm <- gtexCm[,which(
    colnames(gtexCm) %in% gtexSample$SAMPID
)]

# Set benchmark parameters
samplesPrCondition   <- c(5,20,50) 
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

# Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'

    library(doMC)
    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

# list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
# Sanity check
length(nrRepList) == length(samplesPrCondition) * nrRepsMade
table(sapply(nrRepList, length))
```

# Generate leniently filtered Gtex benchmark datasets

```{r}
gtexBenchmarkDataLenient <- lapply(c(1:length(nrRepList)), function(x) {
      
    # Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]

    if(TRUE) {
        localSubset <- sample(
            colnames(gtexCm),
            localSampleSize * 2
        )

        localDesign <- data.frame(
            sample_id = localSubset,
            condition = c(
                rep('a',floor  ( length(localSampleSize) )),
                rep('b',ceiling( length(localSampleSize) ))
            ),
            stringsAsFactors = FALSE
        )
        localDesign <- localDesign[sort.list(localDesign$condition),]
    }
      
    # Step 2: Subset to expressed features using edgeR::filterByExpr
    if(TRUE) {
            
        y <- edgeR::DGEList(counts = gtexCm[,localDesign$sample_id])
        design <- model.matrix(~condition, data=localDesign)
            
        filter <- edgeR::filterByExpr(y,design=design) 
            
        localCm <- y$counts[filter,]
        # Get only multi-isoform genes (after filtering)
        localTx <- txInfo[which(
            txInfo$isoform_id %in% rownames(localCm)),]
            
        tmp <- table(localTx$gene_id)
        tmp <- tmp[which( tmp >= 2)]
        
        localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
        localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
    }

    # Step 3: Extract isoforms to modify
    if(TRUE) {
        genesToModify <- sample(
            x = unique(localTx$gene_id),
            size = round(
                length(unique(localTx$gene_id)) * fracGenesAffected
            )
        )
            
        samplesToModify <- localDesign$sample_id[which(
            localDesign$condition == 'b')]
            
        cm_swapped <- localCm
            
        transcripts_toSwap_current <- c()
        transcripts_swapped_current <- c()
        transcripts_swapped_all <- c()
        transcripts_toSwap_all <- c()

        for (gene in genesToModify) {

            current <- localTx[which(localTx$gene_id==gene),]
            nSwap <- max(2,rbinom(1,nrow(current),1/3))
  
            transcripts_toSwap_current <- sample(
                x = current$isoform_id,
                size = nSwap)
  
            # swap order of txs completely by putting the fist one last
            transcripts_swapped_current <- c(transcripts_toSwap_current[-1],transcripts_toSwap_current[1])
                
            # Add to swapping queue
            transcripts_swapped_all <- c(transcripts_swapped_all,transcripts_swapped_current)
            transcripts_toSwap_all <- c(transcripts_toSwap_all,transcripts_toSwap_current)
        }
        
    # Perform the swapping in matrix
    cm_swapped[transcripts_toSwap_all,which(colnames(cm_swapped)%in%samplesToModify)] <- cm_swapped[transcripts_swapped_all,which(colnames(cm_swapped)%in%samplesToModify)]
      
    # Set swapping in localTx
    localTx$txSwapped <- vector(length=nrow(localTx))
    localTx[which(localTx$isoform_id %in% transcripts_swapped_all),"txSwapped"] <- TRUE 
    localTx$nrSamplesPerCondition <- localSampleSize
    }
    
    colnames(localTx) <- c('TXNAME','GENEID','txSwapped','nrSamplesPerCondition')

    # Combine data
    dataList <- list(
        data     = cm_swapped,
        design   = localDesign,
        metaInfo = localTx
    )

    return(dataList)
})

names(gtexBenchmarkDataLenient) <- paste0(names(nrRepList),"_filterLenient")
```

# Generate stringently filtered Gtex benchmark datasets

```{r}
gtexBenchmarkDataStringent <- lapply(c(1:length(nrRepList)), function(x) {
      
    # Step 1: Extract random sub-sample of correct size
    set.seed(x)
    localSampleSize <- nrRepList[[x]]

    if(TRUE) {
        localSubset <- sample(
            colnames(gtexCm),
            localSampleSize * 2
        )

        localDesign <- data.frame(
            sample_id = localSubset,
            condition = c(
                rep('a',floor  ( length(localSampleSize) )),
                rep('b',ceiling( length(localSampleSize) ))
            ),
            stringsAsFactors = FALSE
        )
        localDesign <- localDesign[sort.list(localDesign$condition),]
    }
      
    # Step 2: Subset to expressed features using DRIMSeq::dmFilter
    if(TRUE) {
          
      localCm <- gtexCm[,localDesign$sample_id]

      geneForEachTx <- txInfo[match(rownames(localCm),txInfo[,"isoform_id"]),"gene_id"]

      localCm$gene_id <- geneForEachTx
      localCm$feature_id <- row.names(localCm)

      d <- DRIMSeq::dmDSdata(counts = localCm, samples = localDesign)

      d_filter <- dmFilter(d,
                min_samps_feature_expr=localSampleSize/2, 
                min_feature_expr=10, 
                min_samps_feature_prop=localSampleSize/2, 
                min_feature_prop=0.1,
                min_samps_gene_expr=localSampleSize, 
                min_gene_expr=10)

      localCm <- gtexCm[counts(d_filter)$feature_id,localDesign$sample_id]

      # Get only multi-isoform genes (after filtering)
      localTx <- txInfo[which(
          txInfo$isoform_id %in% rownames(localCm)),]
          
      tmp <- table(localTx$gene_id)
      tmp <- tmp[which( tmp >= 2)]
        
      localTx <- localTx[which(localTx$gene_id %in% names(tmp)),]
      localCm <- localCm[which(rownames(localCm) %in% localTx$isoform_id),]
    }

    # Step 3: Extract isoforms to modify
    if(TRUE) {
        genesToModify <- sample(
            x = unique(localTx$gene_id),
            size = round(length(unique(localTx$gene_id)) * fracGenesAffected))
            
        samplesToModify <- localDesign$sample_id[which(
            localDesign$condition == 'b')]
            
        cm_swapped <- localCm
            
        transcripts_toSwap_current <- c()
        transcripts_swapped_current <- c()
        transcripts_swapped_all <- c()
        transcripts_toSwap_all <- c()

        for (gene in genesToModify) {

            current <- localTx[which(localTx$gene_id==gene),]
            nSwap <- max(2,rbinom(1,nrow(current),1/3))
  
            transcripts_toSwap_current <- sample(
                x = current$isoform_id,
                size = nSwap)
  
            # swap order of txs completely by putting the fist one last
            transcripts_swapped_current <- c(transcripts_toSwap_current[-1],transcripts_toSwap_current[1])
                
            # Add to swapping queue
            transcripts_swapped_all <- c(transcripts_swapped_all,transcripts_swapped_current)
            transcripts_toSwap_all <- c(transcripts_toSwap_all,transcripts_toSwap_current)
        }
        
    # Perform the swapping in matrix
    cm_swapped[transcripts_toSwap_all,which(colnames(cm_swapped)%in%samplesToModify)] <- cm_swapped[transcripts_swapped_all,which(colnames(cm_swapped)%in%samplesToModify)]
      
    # Set swapping in localTx
    localTx$txSwapped <- vector(length=nrow(localTx))
    localTx[which(localTx$isoform_id %in% transcripts_swapped_all),"txSwapped"] <- TRUE 
    localTx$nrSamplesPerCondition <- localSampleSize
    }
    
    colnames(localTx) <- c('TXNAME','GENEID','txSwapped','nrSamplesPerCondition')

    # Combine data
    dataList <- list(
        data     = cm_swapped,
        design   = localDesign,
        metaInfo = localTx
    )

    return(dataList)
})

names(gtexBenchmarkDataStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save gtex benchmark data

This file is already present at /satuRn_paper/Data/working_data/Gtex_data

```{r}
save(gtexBenchmarkDataLenient, gtexBenchmarkDataStringent, file=paste0(wd,'/Data/working_data/Gtex_data/01_gtex_benchmark_data_count.Rdata'))
```





