---
title: "1_Darmanis_prepare"
author: "Jeroen Gilis"
date: "05/11/2020"
output: html_document
---

**In order to run this script (1_Darmanis_perpare.Rmd), the expression count matrix Darmanis_counts.rds and corresponding metadata Darmanis_metadata.Rdata should be downloaded from Zenodo and put in the Data folder of this GitHub repository.** Alternatively, the exact same quantification files can be downloaded from their original source: http://imlspenticton.uzh.ch/robinson_lab/conquer/data-mae/GSE84465.rds .

**If one does not want to run this script, its output can also be downloaded from Zenodo: Darmanis_benchmark_datasets_count.Rdata (or, alternatively, Darmanis_benchmark_datasets_scaledTPM.Rdata) **

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Load libraries

```{r,message=FALSE,warning=FALSE}
library(edgeR)
library(DRIMSeq)
library(MultiAssayExperiment)
library(doMC)
```

# Load data

```{r, message=FALSE}
Darmanis_data <- readRDS(file = "./Data/Darmanis_counts.rds")
```

# Extract genome annotation

```{r}
### Extract gene info
txInfo <- Darmanis_data@ExperimentList$tx@rowRanges
txInfo <- as.data.frame(mcols(txInfo[,c('transcript','gene')]))
colnames(txInfo) <- c('isoform_id','gene_id')
rownames(txInfo) <- NULL
txInfo$isoform_id <- as.character(txInfo$isoform_id)
txInfo$gene_id <- as.character(txInfo$gene_id)
```

# Subset Darmanis data

To generate a benchmark dataset, we need to know the ground truth, i.e. which transcripts are differentially used between groups of samples. To this end, we first generate mock datasets where no DTU is expected, after which we introduce DTU between groups by the transcript-abundance-swapping strategy discussed by Van den Berge et al. [stageR paper](https://doi.org/10.1186/s13059-017-1277-0).

In order to obtain a mock dataset, we need to construct a dataset where (ideally) no DTU is expected. We try to achieve this by subsampling a homogenous set of samples from the Darmanis dataset, based on following criteria:

1. cell.type = Immune cell
2. tsne.cluster = 8
3. tissue.ch1 = Tumor

The metadata of the retained cells can be retrieved from Darmanis_metadata.Rdata;

```{r}
load(file = "./Data/Darmanis_metadata.Rdata")
### Subset expression matrix
Darmanis_counts <- Darmanis_data@ExperimentList$tx@assays$data$count[,cellsOfInterest] # count input
```

Uncomment this chunk of code to use scaledTPM in stead  of raw counts

```{r}
# # get scaledTPM
# Darmanis_counts_all <- Darmanis_data@ExperimentList$tx@assays$data$count[,cellsOfInterest] # count input
# Darmanis_TPM_all <- Darmanis_data@ExperimentList$tx@assays$data$TPM[,cellsOfInterest] # TPM input
# 
# colsums_count <- colSums(Darmanis_counts_all)
# colsums_TPM <- colSums(Darmanis_TPM_all)
# scale_factors <- colsums_count/colsums_TPM
# 
# Darmanis_scaledTPM_all <- Darmanis_TPM_all %*% diag(scale_factors) # scaledTPM input
# colnames(Darmanis_scaledTPM_all) <- colnames(Darmanis_counts_all)
# Darmanis_counts <- Darmanis_scaledTPM_all[,cellsOfInterest]
```

# Setup generation of benchmark data

Set hyperparameters of the benchmark datasets (i.e. sample size, number of repeats, fraction of DTU genes and parameters for parallel processing).

```{r}
### Remove ERCC
txInfo <- txInfo[which(
    ! grepl('^ERCC-', txInfo$isoform_id )
),]

Darmanis_counts <- Darmanis_counts[txInfo$isoform_id,]
#Darmanis_scaledTPM <- Darmanis_scaledTPM[txInfo$isoform_id,]

### Set parameters
samplesPrCondition   <- c(20,50,100)
nrRepsMade           <- 3
fracGenesAffected    <- 0.15
nrCoresToUse         <- 2

### Set up parallel processing
if(nrCoresToUse != 1) {
    doParallel <- TRUE
    doProgress <- 'none'

    registerDoMC(cores = nrCoresToUse)
} else {
    doParallel <- FALSE
    doProgress <- 'text'
}

### list for looping
nrRepList <- split(
    rep(
        x = samplesPrCondition,
        times = nrRepsMade
    ),
    paste0(
        'samples_used_',
        rep(
            x = samplesPrCondition,
            times = nrRepsMade
        ),
        '_rep_',
        sort( rep(
            x = 1:nrRepsMade,
            times = length(samplesPrCondition)
        ) )
    )
)
```

# Generate Darmanis benchmark data

```{r}
source(file="./Performance_benchmarks/getBenchmark_data.R")
```

```{r}
DarmanisBenchmarkLenient <- getBenchmark_data(countData=Darmanis_counts, 
                                          metaData=txInfo,
                                          filter="edgeR",
                                          edgeR_filter_spec = list(min.count = 1, 
                                                                    min.total.count = 0, 
                                                                    large.n = 0, 
                                                                    min.prop = 0.5),
                                          nrRepList=nrRepList, 
                                          fracGenesAffected=0.15)
names(DarmanisBenchmarkLenient) <- paste0(names(nrRepList),"_filterLenient")

DarmanisBenchmarkStringent <- getBenchmark_data(countData=Darmanis_counts,
                                                metaData=txInfo,
                                                filter="DRIMSeq", 
                                                nrRepList=nrRepList, 
                                                fracGenesAffected=0.15)
names(DarmanisBenchmarkStringent) <- paste0(names(nrRepList),"_filterStringent")
```

# Save Darmanis benchmark data

```{r}
#save(DarmanisBenchmarkLenient, DarmanisBenchmarkStringent, file="./Data/Darmanis_benchmark_datasets_count.Rdata")
save(DarmanisBenchmarkLenient, DarmanisBenchmarkStringent, file="./Data/Darmanis_benchmark_datasets_scaledTPM.Rdata")
```
